### DISCUSSION TITLE
How to explain the GNN network of MessagePassing class

### DISCUSSION THREAD

**Discussion by 688spiderbin** [pyg-team/pytorch_geometric#9074] [Q&A]
Hello!
Could you update the tutorial of Captum for network of MessagePassing class in the doc. I have tried searching some discussions about how to explain the messagepassing class. But it's quite messy and it's hard to follow.  

I appreciate a lot !

**Comment by rusty1s** [comment]
Can you clarify what you want to have updated exactly, and where you are stuck?

**Reply by 688spiderbin** [reply]
Thanks for your reply, I have got a MessagePassing network for graph classification and I don't know how to do node/edge explainability. Since the official colab example is about using Captum to explain class of  nn.Module (not MessagePassing class). They inherit different classes. Thus, I don't know where to start and maybe need a complete demonstration/example. 
Thank you !

**Reply by rusty1s** [reply]
Have you taken a look at
* https://github.com/pyg-team/pytorch_geometric/blob/master/examples/explain/captum_explainer.py
* https://github.com/pyg-team/pytorch_geometric/blob/master/examples/explain/captum_explainer_hetero_link.py

**Reply by 688spiderbin** [reply]
Yes, the example you show here is nn.Module (The red box). How about MessagePassing model (The blue box) ?

![微信截图_20240325202103](https://github.com/pyg-team/pytorch_geometric/assets/62921113/1eee6879-899e-4580-be95-8a7012fe867f)
![微信截图_20240325202037](https://github.com/pyg-team/pytorch_geometric/assets/62921113/646ded31-a968-4b70-bfd6-554b7cd9ad98)
![微信截图_20240325202614](https://github.com/pyg-team/pytorch_geometric/assets/62921113/cca802c8-057c-4e7b-b54a-340eb026337c)

**Reply by rusty1s** [reply]
A GNN usually consists a `MessagePassing` building blocks, so `MessagePassing` layers are defined within a `torch.nn.Module`. Do you want to explain a single GNN layer?

**Reply by 688spiderbin** [reply]
Probably not to explain a single GNN layer. Anyway, I followed the example you gave me and I stuck at the input parameter for explainer.  
I am doing chemical graph binary classification and I want to know which node or edge would have positive/negative impact on output. The model I used is a pre-trained model, which received four parameters (see below). The number of the tensor([[0.0310]]) is the result of "print(pred)".
And now how to put these parameters in explainer function?  
![微信截图_20240401183240](https://github.com/pyg-team/pytorch_geometric/assets/62921113/b8afa610-0bac-44d9-a473-4623a7845ee8)
![微信截图_20240401183307](https://github.com/pyg-team/pytorch_geometric/assets/62921113/8f2e331d-bfe2-47d7-9f6d-9b928128ac00)

Thanks for you answer!

**Reply by rusty1s** [reply]
You need to input additional arguments except for `x` and `edge_index` as keyword arguments:
```python
explanation = explainer(data.x, data.edge_index, edge_attr=data.edge_attr, batch=data.batch)
```
Sorry if this was not clear from the documentation.

**Reply by 688spiderbin** [reply]
Hi, I have tried your suggestion. It seems that it doesn't work. 
![微信截图_20240403191227](https://github.com/pyg-team/pytorch_geometric/assets/62921113/d827fff9-6a0c-42ef-b5d6-15a37604f714)

**Reply by rusty1s** [reply]
How is your model forward defined? Does there exist an `edge_attr` argument?

**Reply by 688spiderbin** [reply]
The forward function is defined by *argv parameter. And again I could used it through:
`model(data.x, data.edge_index, data.edge_attr, data.batch)`

![微信截图_20240404175233](https://github.com/pyg-team/pytorch_geometric/assets/62921113/edc06131-ab86-46f8-91a1-40e40da3ce88)

**Reply by rusty1s** [reply]
`Explainer` doesn't support `*args` as part of model forward definition. I suggest to simply define it properly:
```
def forward(self, x, edge_index, edge_attr, batch)

**Reply by 688spiderbin** [reply]
Thanks, I followed your suggestion and changed the forward function. And now encounter new bug and it's hard to find the error. 

![微信截图_20240415154221](https://github.com/pyg-team/pytorch_geometric/assets/62921113/d3acb34d-7c5e-4482-ba1d-65ee6d87cdb9)

**Reply by rusty1s** [reply]
Which `captum` version are you using? Do you have a minimal reproducible example?

**Reply by 688spiderbin** [reply]
My captum version is 0.7.  Unfortunately I don't have any successful example.

**Reply by rusty1s** [reply]
Can you downgrade to `<0.7`. I don't think we support the latest `captum` yet.

**Reply by 688spiderbin** [reply]
ok, I have change it to 0.6.  And which function could set this parameter "allow_unused=True". 
![微信截图_20240415201612](https://github.com/pyg-team/pytorch_geometric/assets/62921113/a3acb683-c0af-4854-930f-c3e4870f5972)

**Reply by rusty1s** [reply]
Is this for heterogeneous graphs? It might be that a few node types are not reachable during message passing due to limited amount of hops.

**Reply by 688spiderbin** [reply]
Yep it may be. These are molecular graph while the nodes have node_attr, representing its atomic number and chirality. And the edges have edge_attr, representing single bond or double bond. So in this case, what do I need to do

**Reply by rusty1s** [reply]
Let me try to reproduce and see if we can fix this on our end.

**Reply by 688spiderbin** [reply]
I appreciate a lot. The pretrained-model of GNN I used is come from the article "STRATEGIES FOR PRE-TRAINING GRAPH NEURAL
NETWORKS" , ICLR(2020) by **Weihua Hu** et.al. You could check this model at your convenience. And I try to use the the algorithm of  "ShapleyValueSampling" in CaptumExplainer and it could give me the results surprisingly. 

It seems that  other algorithms refer to gradients may fail in these pretrained models.  I found that it would always fail in the code  "grads = torch.autograd.grad(torch.unbind(outputs), inputs)" in gradients.py in line 119.

By the way, I also try other explanation algorithms refer to the documentation such as GNNexplainer, PGEexplainer. Unfortunately I haven't get any successful example.

**Reply by rusty1s** [reply]
Oh, this looks to be not a problem with heterogeneous graphs then. I think what you can do is to find the unused parameters on your own: https://discuss.pytorch.org/t/how-to-find-the-unused-parameters-in-network/63948/3

**Reply by 688spiderbin** [reply]
Ok, I run through this example https://github.com/pyg-team/pytorch_geometric/blob/master/examples/explain/captum_explainer.py successfully and found that the return of **"grads = torch.autograd.grad(torch.unbind(outputs), inputs)"**  is **{tuple:2} (tensor[...], tensor[...])**.

While my case is **{tuple:2} (None, tensor[...])**  when I set **"grads = torch.autograd.grad(torch.unbind(outputs), inputs, allow_unused=True)"**. So how to deal with this unused parameters, should I remove them?

**Reply by 688spiderbin** [reply]
Because of the None type object, then it would fail in "grad.contiguous()"
![微信截图_20240419142330](https://github.com/pyg-team/pytorch_geometric/assets/62921113/11b81e36-99ba-431b-be84-9a023c14411a)

**Reply by rusty1s** [reply]
I testes both `CaptumExplainer` and `GNNExplainer` with unused parameters and they seem to be working fine, see https://github.com/pyg-team/pytorch_geometric/pull/9216. Is there any easy way for me to reproduce this?

**Reply by 688spiderbin** [reply]
em... If not the case of unused parameter, Could it be some other problems?
Unfortunately I am not clear about what situation would cause this error.  
![微信截图_20240419190959](https://github.com/pyg-team/pytorch_geometric/assets/62921113/c8788bbc-f1a6-4663-8b58-4f963264d6c5)

**Reply by rusty1s** [reply]
Can you post me your model? It may be that the output of your model does not even require gradients, which would explain the error.

**Reply by 688spiderbin** [reply]
Thanks, could your give me an email?

**Reply by 688spiderbin** [reply]
Or I give you the link of the model I use. 
https://github.com/snap-stanford/pretrain-gnns/
Please check the file "/chem/finetune_mutag_ptc.py"

**Reply by rusty1s** [reply]
How is your `CaptumExplainer` defined?

**Reply by 688spiderbin** [reply]
I remember I upload that screen shot before, please check : 
![微信截图_20240426200748](https://github.com/pyg-team/pytorch_geometric/assets/62921113/f9c6c511-b5d7-48d5-9cf2-69e107009069)

**Reply by rusty1s** [reply]
Note that you need to pass in additional arguments as keyword arguments in the explainer:
```python
explainer(data.x, data.edge_index, edge_attr=data.edge_attr, batch=data.batch)
```
I am not sure if this resolved your issue.

**Reply by 688spiderbin** [reply]
Yep, you have suggested to me this before, it's cannot work.

**Reply by rusty1s** [reply]
Sorry for late reply. This works fine on our end, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/test/explain/algorithm/test_captum_explainer.py#L149). I am not yet sure what is the problem though.

**Comment by 688spiderbin** [comment]
Thanks for your recommendation, I rechecked the model and can't found any parameter haven't use. Indeed It's hard to detect the issue and describe it exactly, so I paste all the code here.  

```
import torch
from torch_geometric.nn import MessagePassing
from torch_geometric.utils import add_self_loops, degree, softmax
from torch_geometric.nn import global_add_pool, global_mean_pool, global_max_pool, GlobalAttention, Set2Set
import torch.nn.functional as F
from torch_scatter import scatter_add
from torch_geometric.nn.inits import glorot, zeros
```

The model is 
```
class GINConv(MessagePassing):
    def __init__(self, emb_dim, aggr = "add"):
        super(GINConv, self).__init__()
        #multi-layer perceptron
        self.mlp = torch.nn.Sequential(torch.nn.Linear(emb_dim, 2*emb_dim), torch.nn.ReLU(), torch.nn.Linear(2*emb_dim, emb_dim))
        self.edge_embedding1 = torch.nn.Embedding(num_bond_type, emb_dim)
        self.edge_embedding2 = torch.nn.Embedding(num_bond_direction, emb_dim)

        torch.nn.init.xavier_uniform_(self.edge_embedding1.weight.data)
        torch.nn.init.xavier_uniform_(self.edge_embedding2.weight.data)
        self.aggr = aggr

    def forward(self, x, edge_index, edge_attr):
        #add self loops in the edge space
        edge_index = add_self_loops(edge_index, num_nodes = x.size(0))[0]

        #add features corresponding to self-loop edges.
        self_loop_attr = torch.zeros(x.size(0), 2)
        self_loop_attr[:,0] = 4 #bond type for self-loop edge
        self_loop_attr = self_loop_attr.to(edge_attr.device).to(edge_attr.dtype)
        edge_attr = torch.cat((edge_attr, self_loop_attr), dim = 0)

        edge_embeddings = self.edge_embedding1(edge_attr[:,0]) + self.edge_embedding2(edge_attr[:,1])

        # return self.propagate(self.aggr, edge_index, x=x, edge_attr=edge_embeddings)
        return self.propagate(edge_index, x=x, edge_attr=edge_embeddings)

    def message(self, x_j, edge_attr):
        return x_j + edge_attr

    def update(self, aggr_out):
        return self.mlp(aggr_out)
```

```
class GNN(torch.nn.Module):
   """
    Args:
        num_layer (int): the number of GNN layers
        emb_dim (int): dimensionality of embeddings
        JK (str): last, concat, max or sum.
        max_pool_layer (int): the layer from which we use max pool rather than add pool for neighbor aggregation
        drop_ratio (float): dropout rate
        gnn_type: gin, gcn, graphsage, gat

    Output:
        node representations

    """
    def __init__(self, num_layer, emb_dim, JK = "last", drop_ratio = 0, gnn_type = "gin"):
        super(GNN, self).__init__()
        self.num_layer = num_layer
        self.drop_ratio = drop_ratio
        self.JK = JK

        if self.num_layer < 2:
            raise ValueError("Number of GNN layers must be greater than 1.")

        self.x_embedding1 = torch.nn.Embedding(num_atom_type, emb_dim)
        self.x_embedding2 = torch.nn.Embedding(num_chirality_tag, emb_dim)

        torch.nn.init.xavier_uniform_(self.x_embedding1.weight.data)
        torch.nn.init.xavier_uniform_(self.x_embedding2.weight.data)

        ###List of MLPs
        self.gnns = torch.nn.ModuleList()
        for layer in range(num_layer):
            if gnn_type == "gin":
                self.gnns.append(GINConv(emb_dim, aggr = "add"))
            elif gnn_type == "gcn":
                self.gnns.append(GCNConv(emb_dim))
            elif gnn_type == "gat":
                self.gnns.append(GATConv(emb_dim))
            elif gnn_type == "graphsage":
                self.gnns.append(GraphSAGEConv(emb_dim))


        ###List of batchnorms
        self.batch_norms = torch.nn.ModuleList()
        for layer in range(num_layer):
            self.batch_norms.append(torch.nn.BatchNorm1d(emb_dim))

    def forward(self, x, edge_index, edge_attr):
        x = self.x_embedding1(x[:, 0].to(dtype=torch.int64)) + self.x_embedding2(x[:, 1].to(dtype=torch.int64))
        h_list = [x]
        for layer in range(self.num_layer):
            h = self.gnns[layer](h_list[layer], edge_index, edge_attr)
            h = self.batch_norms[layer](h)
            #h = F.dropout(F.relu(h), self.drop_ratio, training = self.training)
            if layer == self.num_layer - 1:
                #remove relu for the last layer
                h = F.dropout(h, self.drop_ratio, training = self.training)
            else:
                h = F.dropout(F.relu(h), self.drop_ratio, training = self.training)
            h_list.append(h)

        ### Different implementations of Jk-concat
        if self.JK == "concat":
            node_representation = torch.cat(h_list, dim = 1)
        elif self.JK == "last":
            node_representation = h_list[-1]
        elif self.JK == "max":
            h_list = [h.unsqueeze_(0) for h in h_list]
            node_representation = torch.max(torch.cat(h_list, dim = 0), dim = 0)[0]
        elif self.JK == "sum":
            h_list = [h.unsqueeze_(0) for h in h_list]
            node_representation = torch.sum(torch.cat(h_list, dim = 0), dim = 0)[0]

        return node_representation
```
```
class GNN_graphpred(torch.nn.Module):
    """
    Extension of GIN to incorporate edge information by concatenation.

    Args:
        num_layer (int): the number of GNN layers
        emb_dim (int): dimensionality of embeddings
        num_tasks (int): number of tasks in multi-task learning scenario
        drop_ratio (float): dropout rate
        JK (str): last, concat, max or sum.
        graph_pooling (str): sum, mean, max, attention, set2set
        gnn_type: gin, gcn, graphsage, gat
        
    See https://arxiv.org/abs/1810.00826
    JK-net: https://arxiv.org/abs/1806.03536
    """
    def __init__(self, num_layer, emb_dim, num_tasks, gnn_type, JK = "last", drop_ratio = 0, graph_pooling = "mean"):
        super(GNN_graphpred, self).__init__()
        self.num_layer = num_layer
        self.drop_ratio = drop_ratio
        self.JK = JK
        self.emb_dim = emb_dim
        self.num_tasks = num_tasks

        if self.num_layer < 2:
            raise ValueError("Number of GNN layers must be greater than 1.")

        self.gnn = GNN(num_layer, emb_dim, JK, drop_ratio, gnn_type = gnn_type)

        #Different kind of graph pooling
        if graph_pooling == "sum":
            self.pool = global_add_pool
        elif graph_pooling == "mean":
            self.pool = global_mean_pool
        elif graph_pooling == "max":
            self.pool = global_max_pool
        elif graph_pooling == "attention":
            if self.JK == "concat":
                self.pool = GlobalAttention(gate_nn = torch.nn.Linear((self.num_layer + 1) * emb_dim, 1))
            else:
                self.pool = GlobalAttention(gate_nn = torch.nn.Linear(emb_dim, 1))
        elif graph_pooling[:-1] == "set2set":
            set2set_iter = int(graph_pooling[-1])
            if self.JK == "concat":
                self.pool = Set2Set((self.num_layer + 1) * emb_dim, set2set_iter)
            else:
                self.pool = Set2Set(emb_dim, set2set_iter)
        else:
            raise ValueError("Invalid graph pooling type.")

        #For graph-level binary classification
        if graph_pooling[:-1] == "set2set":
            self.mult = 2
        else:
            self.mult = 1
        
        if self.JK == "concat":
            self.graph_pred_linear = torch.nn.Linear(self.mult * (self.num_layer + 1) * self.emb_dim, self.num_tasks)
        else:
            self.graph_pred_linear = torch.nn.Linear(self.mult * self.emb_dim, self.num_tasks)

    def from_pretrained(self, model_file):
        #self.gnn = GNN(self.num_layer, self.emb_dim, JK = self.JK, drop_ratio = self.drop_ratio)
        self.gnn.load_state_dict(torch.load(model_file, map_location='cpu'))

    def forward(self, x, edge_index, edge_attr, batch):
        node_representation = self.gnn(x, edge_index, edge_attr)
        return self.graph_pred_linear(self.pool(node_representation, batch)) 
```

And the train function is 
```
criterion = nn.BCEWithLogitsLoss(reduction = "none")

def train(args, model, device, loader, optimizer):
    model.train()

    for step, batch in enumerate(tqdm(loader, desc="Iteration")):
        batch = batch.to(device)
        pred = model(batch.x, batch.edge_index, batch.edge_attr, batch.batch)
        y = batch.y.view(pred.shape).to(torch.float64)

        #Whether y is non-null or not.
        is_valid = y**2 > 0
        #Loss matrix
        loss_mat = criterion(pred.double(), (y+1)/2)
        #loss matrix after removing null target
        loss_mat = torch.where(is_valid, loss_mat, torch.zeros(loss_mat.shape).to(loss_mat.device).to(loss_mat.dtype))
            
        optimizer.zero_grad()
        loss = torch.sum(loss_mat)/torch.sum(is_valid)
        loss.backward()
        optimizer.step()
```

The code for explanation 
```
# set-up dataset
dataset = MoleculeDataset("E:/github_download/SourceCode/pretrain-gnns-master/dataset/" + 'mutag', dataset='mutag')

train_dataset, valid_dataset = cv_random_split(dataset, seed=42)
print(train_dataset[0])
data = ([t for t in train_dataset])[0]  # choose one 
print(data)


# using pre-trained model
input_model_file = "E:/github_download/SourceCode/pretrain-gnns-master/chem/model_gin/supervised_masking.pth"
model = GNN_graphpred(5, 300, 1, gnn_type='gin', JK='last',
                      drop_ratio=0.5,
                      graph_pooling="mean")
model.from_pretrained(input_model_file)


model.to(device)
model.eval()
data = data.to(device)
pred = model(data.x, data.edge_index, data.edge_attr, data.batch)
print(pred)


explainer = Explainer(
    model = model,
    algorithm=CaptumExplainer('IntegratedGradients'),  
    explanation_type='model',
    model_config=dict(
        mode = 'binary_classification',
        task_level = 'graph',
        return_type = 'probs',
    ),
    node_mask_type='attributes',
    edge_mask_type='object',
    threshold_config=dict(
        threshold_type='topk',
        value=200,
    ),
)


explanation = explainer(data.x, data.edge_index, edge_attr=data.edge_attr, batch=data.batch)
print(explanation.available_explanations)

path = 'feature_importance.png'
explanation.visualize_feature_importance(path, top_k=10)
print(f"Feature importance plot has been saved to '{path}'")

path = 'subgraph.pdf'
explanation.visualize_graph(path)
print(f"Subgraph plot has been saved to '{path}'")
```

Then, it comes up
```
Traceback (most recent call last):
  File "E:\github_download\SourceCode\pretrain-gnns-master\chem\explain(成功).py", line 64, in <module>
    explanation = explainer(data.x, data.edge_index, edge_attr=data.edge_attr, batch=data.batch)
  File "E:\anaconda3\envs\py38\lib\site-packages\torch_geometric\explain\explainer.py", line 205, in __call__
    explanation = self.algorithm(
  File "E:\anaconda3\envs\py38\lib\site-packages\torch\nn\modules\module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\anaconda3\envs\py38\lib\site-packages\torch_geometric\explain\algorithm\captum_explainer.py", line 170, in forward
    attributions = self.attribution_method_instance.attribute(
  File "E:\anaconda3\envs\py38\lib\site-packages\captum\log\__init__.py", line 42, in wrapper
    return func(*args, **kwargs)
  File "E:\anaconda3\envs\py38\lib\site-packages\captum\attr\_core\integrated_gradients.py", line 274, in attribute
    attributions = _batch_attribution(
  File "E:\anaconda3\envs\py38\lib\site-packages\captum\attr\_utils\batching.py", line 78, in _batch_attribution
    current_attr = attr_method._attribute(
  File "E:\anaconda3\envs\py38\lib\site-packages\captum\attr\_core\integrated_gradients.py", line 351, in _attribute
    grads = self.gradient_func(           # 在这里计算梯度
  File "E:\anaconda3\envs\py38\lib\site-packages\captum\_utils\gradient.py", line 119, in compute_gradients
    grads = torch.autograd.grad(torch.unbind(outputs), inputs)
  File "E:\anaconda3\envs\py38\lib\site-packages\torch\autograd\__init__.py", line 303, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.
```

**Reply by 688spiderbin** [reply]
AND Could you help me to find out which parameters haven't used according to the code? Is it torch.nn.Embedding ? I don't think this will be the answer. 

```
>>>self.x_embedding1(x[:, 0].to(dtype=torch.int64)) + self.x_embedding2(x[:, 1].to(dtype=torch.int64))
Out[1]: 
tensor([[ 1.4876, -0.7566,  3.8224,  ..., -3.0525, -2.4919, -1.2139],
        [ 1.4876, -0.7566,  3.8224,  ..., -3.0525, -2.4919, -1.2139],
        [ 1.4876, -0.7566,  3.8224,  ..., -3.0525, -2.4919, -1.2139],
        ...,
        [ 0.1232, -0.7887,  3.5018,  ...,  3.5915,  1.4795, -1.6173],
        [-2.2018,  0.8337,  3.6481,  ...,  2.3498, -0.4928,  0.4342],
        [-2.2018,  0.8337,  3.6481,  ...,  2.3498, -0.4928,  0.4342]],
       grad_fn=<AddBackward0>)
```
